src/driver/op.rs

```diff
@@ -183,6 +194,78 @@ where
+        use std::mem;
+
+        let me = &mut *self;
+
+        CONTEXT.with(|runtime_context| {
+            runtime_context.with_driver_mut(|driver| {
+                let (lifecycle, completions) = driver
+                    .ops
+                    .get_mut(me.index)
+                    .expect("invalid internal state");
+
+                match mem::replace(lifecycle, Lifecycle::Submitted) {
+                    Lifecycle::Submitted => {
+                        *lifecycle = Lifecycle::Waiting(cx.waker().clone());
+                        Poll::Pending
+                    }
+                    Lifecycle::Waiting(waker) if !waker.will_wake(cx.waker()) => {
+                        *lifecycle = Lifecycle::Waiting(cx.waker().clone());
+                        Poll::Pending
+                    }
+                    Lifecycle::Waiting(waker) => {
+                        *lifecycle = Lifecycle::Waiting(waker);
+                        Poll::Pending
+                    }
+                    Lifecycle::Ignored(..) => unreachable!(),
+                    Lifecycle::Completed(cqe) => {
+                        // This is possible. We may have previously polled a CompletionList,
+                        // and the final CQE registered as Completed
+                        driver.ops.remove(me.index);
+                        me.index = usize::MAX;
+                        Poll::Ready(me.data.take().unwrap().complete(cqe))
+                    }
+                    Lifecycle::CompletionList(indices) => {
+                        let mut data = me.data.take().unwrap();
+                        let mut status = Poll::Pending;
+                        // Consume the CqeResult list, calling update on the Op on all Cqe's flagged `more`
+                        // If the final Cqe is present, clean up and return Poll::Ready
+                        for cqe in indices.into_list(completions) {
+                            if io_uring::cqueue::more(cqe.flags) {
+                                data.update(cqe);
+                            } else {
+                                status = Poll::Ready(cqe);
+                                break;
+                            }
+                        }
+                        match status {
+                            Poll::Pending => {
+                                // We need more CQE's. Restore the op state
+                                let _ = me.data.insert(data);
+                                *lifecycle = Lifecycle::Waiting(cx.waker().clone());
+                                Poll::Pending
+                            }
+                            Poll::Ready(cqe) => {
+                                me.index = usize::MAX;
+                                driver.ops.remove(me.index);
+                                Poll::Ready(data.complete(cqe))
+                            }
+                        }
```

Comment:
This is pulled out of the for loop to allow access to driver.ops, which is otherwise prevented by the borrow checker by the borrow of lifecycle.